### PR TITLE
apiserver: start only one compactor per unique storagebackend transport config

### PR DESCRIPTION
**Pull request by sttts**
Before this PR we were running one compactor per storage (this includes one per registry instance and one per CRD, another one on each CRD change!).

```release-note
Run one etcd storage compaction per default interval of 5min. Do not run one for each resource and each CRD. This fixes the compaction log spam and reduces load on etcd.
```

### CODE REVIEW

**Comment by sttts** [comment]
@joelegasse does this sound reasonable to start only one compactor? What is the load today with 100ish many compactors?

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -293,8 +293,8 @@ func CreateKubeAPIServerConfig(
 		return
 	}
 
-	if _, port, err := net.SplitHostPort(s.Etcd.StorageConfig.ServerList[0]); err == nil && port != "0" && len(port) != 0 {
-		if err := utilwait.PollImmediate(etcdRetryInterval, etcdRetryLimit*etcdRetryInterval, preflight.EtcdConnection{ServerList: s.Etcd.StorageConfig.ServerList}.CheckEtcdServers); err != nil {
+	if _, port, err := net.SplitHostPort(s.Etcd.StorageConfig.Transport.ServerList[0]); err == nil && port != "0" && len(port) != 0 {
```
Comment (yue9944882) [inline cmd/kube-apiserver/app/server.go]: NPE here?

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 2]
```diff
@@ -108,15 +109,81 @@ func newETCD3Client(c storagebackend.Config) (*clientv3.Client, error) {
...
+)
+
+// startCompactorOnce start one compactor per transport. If the interval get smaller on repeated calls, the
+// compactor is replaced. A destroy func is returned. If all destroy funcs with the same transport are called,
+// the compactor is stopped.
+func startCompactorOnce(c storagebackend.TransportConfig, interval time.Duration) (func(), error) {
+	lock.Lock()
+	defer lock.Unlock()
+
+	key := fmt.Sprintf("%v", c)
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: but it looks slightly odd me. `key := fmt.Sprintf("%v", c.ServerList)` cleaner?

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 3]
```diff
@@ -108,15 +109,81 @@ func newETCD3Client(c storagebackend.Config) (*clientv3.Client, error) {
...
+	return func() {
+		lock.Lock()
+		defer lock.Unlock()
+
+		compactor := compactors[key]
+		compactor.refs--
+		if compactors[key].refs == 0 {
+			compactor.cancel()
+			compactor.client.Close()
+			delete(compactors, key)
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: deleting by key here this looks dangerous. we may delete wrong storage on racing by mistake amiright? deleting by pointer equality would be safer.

**Comment by lavalamp** [comment]
We probably also shouldn't run one per apiserver in an HA setup...

Better yet, can we just delete this now? https://github.com/kubernetes/kubernetes/issues/46705#issuecomment-414276152

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (sttts) [reply cmd/kube-apiserver/app/server.go]: Why a NPE? the server list length is validated in the options to be non-zero.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 2 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: no, I want the keys to be part of this as the etcd client depends on those too.

Added a comment.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 3 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: not sure I can follow. The compactor is not by storage, but by transport config. Moreover, we lock before reading or writing to `compactors`.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 4]
```diff
@@ -108,15 +109,81 @@ func newETCD3Client(c storagebackend.Config) (*clientv3.Client, error) {
...
+	client   *clientv3.Client
+	refs     int
+}
+
+var (
+	lock       sync.Mutex
+	compactors = map[string]*runningCompactor{}
+)
+
+// startCompactorOnce start one compactor per transport. If the interval get smaller on repeated calls, the
```
[the remark this answers is not in the document]
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: Our storage API allows different intervals. So we should support it here in the backend as well.

**Comment by sttts** [comment]
@hexfusion could you take a look?

**Comment by sttts** [comment]
@yue9944882 updated. Ptal.

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go [hunk 5]
```diff
@@ -108,16 +109,89 @@ func newETCD3Client(c storagebackend.Config) (*clientv3.Client, error) {
...
+
+		compactor.interval = interval
+		compactor.cancel = cancel
+		compactor.client = compactorClient
+		compactor.refs++
+		compactors[key] = compactor
+
+		etcd3.StartCompactor(ctx, compactorClient, interval)
+	} else {
+		compactor.refs++
```
Comment (yue9944882) [inline staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: nits: move `compactor.refs++` out of if-else block

**Review by yue9944882 — COMMENTED** [review]
nits and suspicious nil panic, otherwise lgtm

File: cmd/kube-apiserver/app/server.go (hunk 1 above)
Comment (yue9944882) [reply cmd/kube-apiserver/app/server.go]: mmm.. yeah, no chance of NPE here

File: staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go (hunk 5 above)
Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: :+1:

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/storage/storagebackend/factory/etcd3.go]: fixed

**Pull request merged**